src/lattice.rs

```diff
@@ -0,0 +1,228 @@
+    ///
+    /// When provided an empty frontier `self` is not modified.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use timely::PartialOrder;
+    /// # use timely::order::Product;
+    /// # use dbsp::lattice::Lattice;
+    /// # fn main() {
+    ///
+    /// use timely::progress::frontier::{Antichain, AntichainRef};
+    ///
+    /// let time = Product::new(3, 7);
+    /// let mut advanced = Product::new(3, 7);
+    /// let frontier = Antichain::from(vec![Product::new(4, 8), Product::new(5, 3)]);
+    /// advanced.advance_by(frontier.borrow());
+    ///
+    /// // `time` and `advanced` are indistinguishable to elements >= an element of `frontier`
+    /// for i in 0..10 {
+    ///     for j in 0..10 {
+    ///         let test = Product::new(i, j);
+    ///         // for `test` in the future of `frontier` ..
+    ///         if frontier.less_equal(&test) {
+    ///             assert_eq!(time.less_equal(&test), advanced.less_equal(&test));
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// assert_eq!(advanced, Product::new(4, 7));
+    /// # }
+    /// ```
+    #[inline]
+    fn advance_by(&mut self, frontier: AntichainRef<Self>)
+    where
+        Self: Sized,
+    {
+        let mut iter = frontier.iter();
+        if let Some(first) = iter.next() {
+            let mut result = self.join(first);
+            for f in iter {
+                result.meet_assign(&self.join(f));
+            }
+            *self = result;
+        }
+    }
+}
+
+use timely::order::Product;
+
+impl<T1: Lattice, T2: Lattice> Lattice for Product<T1, T2> {
+    #[inline]
+    fn join(&self, other: &Product<T1, T2>) -> Product<T1, T2> {
+        Product {
+            outer: self.outer.join(&other.outer),
+            inner: self.inner.join(&other.inner),
+        }
+    }
+    #[inline]
+    fn meet(&self, other: &Product<T1, T2>) -> Product<T1, T2> {
+        Product {
+            outer: self.outer.meet(&other.outer),
+            inner: self.inner.meet(&other.inner),
+        }
+    }
+}
+
+macro_rules! implement_lattice {
+    ($index_type:ty, $minimum:expr) => {
+        impl Lattice for $index_type {
+            #[inline]
+            fn join(&self, other: &Self) -> Self {
+                ::std::cmp::max(*self, *other)
+            }
+            #[inline]
+            fn meet(&self, other: &Self) -> Self {
+                ::std::cmp::min(*self, *other)
+            }
+        }
+    };
+}
+
+use std::time::Duration;
+
+implement_lattice!(Duration, Duration::new(0, 0));
+implement_lattice!(usize, 0);
+implement_lattice!(u128, 0);
+implement_lattice!(u64, 0);
+implement_lattice!(u32, 0);
+implement_lattice!(u16, 0);
+implement_lattice!(u8, 0);
+implement_lattice!(isize, 0);
+implement_lattice!(i128, 0);
+implement_lattice!(i64, 0);
+implement_lattice!(i32, 0);
+implement_lattice!(i16, 0);
+implement_lattice!(i8, 0);
+implement_lattice!((), ());
+
+impl<T: Lattice + Clone> Lattice for Antichain<T> {
```

Comment:
I stole this entire module from DD. If they didn't bother with these optimizations (or didn't want them for some reason), neither will I :)

Reply:
I mean, it's probably just an oversight then, there's no real reason *not* to implement `.*_assign()` here since it allows us to immediately reuse the buffer underlying the `Antichain`